src/hotspot/share/gc/parallel/psPromotionManager.cpp

```diff
@@ -239,16 +240,10 @@ void PSPromotionManager::drain_stacks_depth(bool totally_drain) {
       process_popped_location_depth(task);
     }
 
-    if (totally_drain) {
-      while (tq->pop_local(task)) {
-        process_popped_location_depth(task);
-      }
-    } else {
-      while (tq->size() > _target_stack_size && tq->pop_local(task)) {
-        process_popped_location_depth(task);
-      }
+    while (tq->pop_local(task, threshold)) {
+      process_popped_location_depth(task);
     }
-  } while ((totally_drain && !tq->taskqueue_empty()) || !tq->overflow_empty());
+  } while (!tq->overflow_empty());
```

Comment:
The other, similar loops like the ones in G1 (but also the ones for parallel full gc) all try to first feed the public queue from the overflow queue like (to avoid other threads from starving while this thread happily processes the overflow queue contents on its own):

```
do {
  while (tq->pop_overflow(task)) {
    if (!tq->try_push_to_taskqueue(task)) {
      process_popped_location_depth(task);
    }
  }

  while (tq->pop_local(task, threshold)) {
    process_popped_location_depth(task);
  }
} while (!tq->overflow_empty());
```
Is there a reason to not change the structure of this loop to that better one?

(I am also kind of surprised that this change has not been made yet)